TUI: detect background color

joshtriplett opened the pull request:
(Also reported as a [vim issue](https://github.com/vim/vim/issues/869), though I have more hope of a fix here.)

Vim's automatic background color detection works in GUI mode, but not in a terminal.

On any xterm-compatible terminal, vim could detect the background color via terminal escapes (documented in http://invisible-island.net/xterm/ctlseqs/ctlseqs.html) and use that to automatically detect whether to use `bg=dark` or `bg=light` (by default or with `:set bg&`).

To detect the background color:
- Include `\e]11;?\a' during terminal initialization.
- Asynchronously, when processing input, watch for the escape sequence `\e]11;COLOR\a` (or equivalent using OSC instead of `\e]` for the escape, and ST or `\e\\` instead of `\a` for the terminator, which the existing terminal escape sequence handling probably already handles).  If the COLOR portion matches `rgb:RRRR/GGGG/BBBB` where R, G, and B are hex digits, then [compute the luminance](https://en.wikipedia.org/wiki/Luma_%28video%29) of the RGB color and classify it as light/dark accordingly.  Note that the color components may have anywhere from one to four hex digits, and require scaling accordingly as values out of 4, 8, 12, or 16 bits.

joshtriplett commented:
Apparently vim already includes support for this, though neovim does not.  Search for `rbg_status` in vim to find the implementation.

It looks like the terminal input handling drastically changed in neovim, making this hard to directly port over.  I attached a WIP commit providing an implementation of this for neovim.  It almost works: the background detection works perfectly, as does the setting of the default (so `set bg&` picks up the right default based on the terminal background), but the call to `set_option_value` to set the current background sometimes (but not always) causes a SIGABRT deep in `init_highlight`, while processing `/usr/share/nvim/runtime/syntax/syncolor.vim`.  I haven't yet figured out why, and I'd appreciate some help debugging this.

justinmk commented:
The TUI runs on a separate thread. Most likely that is the cause of the segfault.

justinmk commented on src/nvim/tui/input.c:
```diff
@@ -320,6 +321,60 @@ static bool handle_forced_escape(TermInput *input)
+      if (!option_was_set((char_u *)"bg")) {
+        set_option_value((char_u *)"bg", 0, (char_u *)bgvalue, 0);
+      }
+    }
```
Try putting this block in a function e.g. `set_bg_color_deferred`, then schedule it on the main thread [like this](https://github.com/neovim/neovim/pull/5027/files#diff-9fe05c714b192fe85bb903baf3355196R401)

    loop_schedule(&main_loop, event_create(1, input_done_event, 0));

joshtriplett replied:
Sure enough, that was the problem.  New version momentarily.

joshtriplett committed:
Automatically detect terminal background and set bg=dark or bg=light

xterm-compatible terminals support reporting their configured colors
back to the application.  Use this to obtain the current background
color, compute its luminance to classify it as light or dark, and set
'bg' accordingly.  Also set the default for 'bg', so that `:set bg&`
will revert to that detected default.

joshtriplett commented:
Just pushed a new version, which now works reliably thanks to @justinmk's advice about running on the main loop.

justinmk commented:
LGTM. Is there a termcode to _set_ the background? It could be used to `feed("\027[...")` in `tui_spec.lua`, to set up a test. Otherwise I can't think of a way to test it, it might be too much trouble. But tests help ASan and coverity.

joshtriplett commented:
> Is there a termcode to set the background?

Yes, the same code the terminal writes back to the application: `\x1b]rgb:RRRR/GGGG/BBBB\a`.  However, note that this would change the background of the running terminal.  (And the test itself would only work in a terminal, so it might not work on a CI bot.)

> It could be used to feed("\027[...") in tui_spec.lua, to set up a test. Otherwise I can't think of a way to test it, it might be too much trouble. But tests help ASan and coverity.

Looking at tui_spec.lua, it seems possible to add tests for this: just feed the terminal the escape sequences for various colors, followed by `:set bg`, and see if it has the right background color.

joshtriplett committed:
Add tests for terminal background detection

joshtriplett commented:
I've written tests, and the next push will include them.

I've also fixed the various lint warnings.  Most of them seemed reasonable, but this one seems wrong:
```
src/nvim/tui/input.c:375:  Missing space before asterisk in 0.299*  [whitespace/operators] [2]
```
This line:
```
      double luminance = 0.299*r + 0.587*g + 0.114*b; // CCIR 601
```
seems clearer with the whitespace providing grouping, and less clear with spaces around the `*`s.  Nonetheless, I've changed this to stop lint from failing.

justinmk commented:
typically we would make the grouping explicit with parens.

justinmk commented:
nicely done. I assume the test does fail correctly without your change?

LGTM.

joshtriplett commented:
@justinmk Yes, all the tests fail without the change.

joshtriplett commented:
I'm not sure why varying subsets of the tests fail on some platforms, while some of the tests pass.  I don't know what would cause spurious test failures.

joshtriplett commented:
One interesting caveat about this change: setting the background causes a redraw, which dismisses the vim intro message.  So if you run on a terminal that supports background color detection, and you don't already have `set shortmess+=I`, this will cause that message to immediately disappear.

justinmk commented:
Huh. Does that happen in Vim?

oni-link commented on src/nvim/tui/input.c:
```diff
@@ -320,6 +321,68 @@ static bool handle_forced_escape(TermInput *input)
+      } else if (isxdigit(c)) {
+        if (component < 3 && rgb_max[component] != 0xffff) {
+          uint16_t digit = (uint16_t)((c >= '0' && c <= '9')
+                                      ? c - '0' : 10 + tolower(c) - 'a');
```
`hex2nr()` in `charset.h` could be used here.

joshtriplett replied:
Thanks; done.

oni-link commented on src/nvim/tui/input.c:
```diff
@@ -320,6 +322,66 @@ static bool handle_forced_escape(TermInput *input)
+      } else if (bad) {
+        // ignore
+      } else if (c == '/') {
+        if (component < 3) {
```
Is the sequence not already bad if we see three `'/'`? Should `bad` be set?

justinmk commented:
@joshtriplett Any comment on @oni-link's question? And could you rebase this?

justinmk commented:
Merged in https://github.com/neovim/neovim/pull/9509
